module/mempool/stdmap/eject.go

```diff
@@ -14,41 +14,114 @@ import (
 // backend when it overflows its limit. A custom eject function can be injected
 // into the memory pool upon creation, which allows us to hook into the eject
 // to clean up auxiliary data and/or to change the strategy of eviction.
-type EjectFunc func(entities map[flow.Identifier]flow.Entity) (flow.Identifier, flow.Entity)
+type EjectFunc func(b *Backend) (flow.Identifier, flow.Entity, bool)
 
 // EjectFakeRandom relies on the random map iteration in Go to pick the entity we eject
 // from the entity set. It picks the first entity upon iteration, thus being the fastest
 // way to pick an entity to be evicted; at the same time, it conserves the random bias
 // of the Go map iteration.
-func EjectFakeRandom(entities map[flow.Identifier]flow.Entity) (flow.Identifier, flow.Entity) {
+func EjectFakeRandom(b *Backend) (flow.Identifier, flow.Entity, bool) {
 	var entityID flow.Identifier
 	var entity flow.Entity
-	for entityID, entity = range entities {
+	var bFound bool = false
+	for entityID, entity = range b.entities {
+		bFound = true
 		break
 	}
-	return entityID, entity
+	return entityID, entity, bFound
 }
 
 // EjectTrueRandom relies on a random generator to pick a random entity to eject from the
 // entity set. It will, on average, iterate through half the entities of the set. However,
 // it provides us with a truly evenly distributed random selection.
-func EjectTrueRandom(entities map[flow.Identifier]flow.Entity) (flow.Identifier, flow.Entity) {
+func EjectTrueRandom(b *Backend) (flow.Identifier, flow.Entity, bool) {
 	var entityID flow.Identifier
 	var entity flow.Entity
+	var bFound bool = false
 	i := 0
-	n := rand.Intn(len(entities))
-	for entityID, entity = range entities {
+	n := rand.Intn(len(b.entities))
+	for entityID, entity = range b.entities {
 		if i == n {
+			bFound = true
 			break
 		}
 		i++
 	}
-	return entityID, entity
+	return entityID, entity, bFound
+}
+
+// EjectTrueRandomFast checks if the map size is beyond the
+// ideal size, and will iterate through them and eject unneeded
+// entries if that is the case.
+func EjectTrueRandomFast(b *Backend) (flow.Identifier, flow.Entity, bool) {
+	// 64 in batch plus a buffer of 64 to prevent boundary conditions from
+	// rounding errors
+	const threshold = 128
+
+	const batchSize = 64
+
+	var entities = b.entities
+
+	var retval flow.Identifier 	// dummy value 
+
+	mapSize := len(entities)
+
+	if b.limit > uint(mapSize) {
+		return retval, nil, false
+	}
+
+	if (uint(mapSize) - b.limit) <= threshold {
+		// nothing to do, yet
+		return retval, nil, false
+	}
```

Comment:
This is a minor thing, but to me it's less surprising behaviour if the ejector tries to respect the limit of the backend regardless of batch size. If I call the ejector, I'd expect the mempool to always have size <= limit after that call. In other words the threshold for performing an ejection is always <= the limit of the backend.

Reply:
In a meeting, the backend limit was considered arbitrary, so I'd rather leave this code as is.
Maybe it would be less risky to rename the 'b.limit' to 'b.ejectionTrigger' to reflect its purpose?

Reply:
> less risky to rename the 'b.limit' to 'b.ejectionTrigger' to reflect its purpose?

👍  yea great idea - that's much clearer

Reply:
done